### ISSUE TITLE
Opening a file opens in a new window

### ISSUE THREAD

**Issue by thrashdin** [atom/atom#19147]
### Prerequisites

* [X ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Pressing Ctrl-O now after selecting file always opens it in a new window, not matter what. Same applies to opening a project folder. This happened after last update, but I can't be sure if immediately. Resetting to defaults didn't work. 
Atom    : 1.36.0
Electron: 2.0.18
Chrome  : 61.0.3163.100
Node    : 8.9.3

**Comment by smashwilson** [comment]
1. What operating system are you using?
2. Can you confirm that you're using the literal <kbd>Ctrl-O</kbd> shortcut, as opposed to running `application:open-file` in the command palette, or using the "File -> Open File..." menu option?

**Comment by stepel** [comment]
I have the same problem.
Win 10 PRO.
When I open file from FTP or from local drive every file is opening in the new window instead in the new tab.

I've reinstalled atom to version 1.35 and it works OK.

**Comment by Sokolov57** [comment]
Debian 4.19.28-2 (2019-03-15)
I can confirm this as well as the fact that using either of those opens a new window.

**Comment by politsin** [comment]
> When I open file from FTP or from local drive every file is opening in the new window instead in the new tab.

The same problem with WinSCP

**Comment by stepel** [comment]
> 
> 
> > When I open file from FTP or from local drive every file is opening in the new window instead in the new tab.
> 
> The same problem with WinSCP

Exactly I use WinSCP too but I think it's general problem. 

Another thing what I have noticed is that treeview stopped to open.

**Comment by dima-y** [comment]
I am running Windows x64, and a similar thing started to occur after upgrading to 1.36 - opening a file from a Windows file manager (I'm using Total Commander but don't think it matters) creates a new Atom window.

**Comment by Siguza** [comment]
Same issue on macOS 10.12.6, everything opens in its own window now.

**Comment by dima-y** [comment]
This also occurs in v1.37.0-beta0. Reverting to v1.35.1 fixed it.

**Comment by smashwilson** [comment]
Oh, I see what's going on. It is a similar situation to what people are encountering [on Discourse](https://discuss.atom.io/t/atom-opens-new-files-in-new-window-instead-of-tab/64347/11).

Basically: we changed Atom's launch behavior when started with a single file to not implicitly add the containing directory as a project root. Adding a project root means that Atom recursively indexes everything beneath the file for the fuzzy finder, filesystem watching, and so on - so if you open a file in your home directory, start-up performance goes good-bye. But now, when you launch Atom that way, it has no way to know which Atom window the _next_ file should belong to, because we use the project root directories to determine that, and the single-file window doesn't have any.

For now, if you intend to use an Atom window to work within a project directory on 1.36.0, you can either:

* Open the _directory_ instead of the file directly from Explorer; or
* Use "Add project folder..." to add the directory as a project root once the window is open.

I'm also open to ideas on ways to improve this experience in the near term. I can think of a few:

1. If an Atom window is already open with no project root folders, open new files in that Atom window. (But what do we do if there are multiple Atom windows with no project roots? What should we do when opening new folders?)
2. Provide an easy way to open the folder that contains the file. (But how do we make that path discoverable, and clearly communicate the implications of doing so? What folder do we open if the window already has multiple editors open?)

**Comment by caleb531** [comment]
@smashwilson I think the opened single file should open in the last focused Atom window, a behavior which could be toggled through a new config setting (maybe `core.openSingleFilesInCurrentWindow` or something). I would understand if you wanted to have this setting disabled by default, though IMHO the pre-1.36 behavior is the more intuitive behavior.

If my inspection is correct, I found the relevant logic on Line 989 of `atom-application.js`:
https://github.com/atom/atom/blob/master/src/main-process/atom-application.js#L989

```js
if (!existingWindow && !newWindow && addToLastWindow) {
  let lastWindow = window || this.getLastFocusedWindow()
  if (lastWindow && lastWindow.devMode === devMode) {
    existingWindow = lastWindow
  }
}
```

Perhaps this logic could be tweaked to factor in this proposed config setting?

P.S. I know there's `atom --add` on the command line to trigger this behavior, but most of the single files I'm opening are from the Finder (or from the awesome [Alfred](https://www.alfredapp.com/) app), so this isn't a sufficient alternative.

**Comment by smashwilson** [comment]
"Always open paths in the last window" feels like way too drastic of a change to me. You've seen the fallout from the relatively small change of "don't open a directory when you open a file" here - and _that's_ a change that people have been explicitly asking for for five years; see #1847, #10223, #8934, #18646, and probably others.

> though IMHO the pre-1.36 behavior is the more intuitive behavior.

Users have been confused by our launch behavior for years: see #10517 for an extended discussion. Everyone has an opinion about what the most intuitive launch behavior would be, and everyone's opinions are different. I gave it a shot myself in #18608, and _everyone_ hated it, so I had to revert 😆

I really hesitate to make any further changes without collecting more data about it. Adding a configuration option may help if there simply isn't a consensus of user expectations - but that's just punting the problem, and we still need to decide on a default, and ways of describing each option in a way that users will understand.

We definitely need to address the immediate pain here. But I'd prefer to do so in a way that is as minimally invasive to the rest of our launch behavior as possible.

> P.S. I know there's atom --add on the command line to trigger this behavior, but most of the single files I'm opening are from the Finder (or from the awesome Alfred app), so this isn't a sufficient alternative.

I wonder if there's a way for us to behave differently when launched from shell integration, or even offer multiple entry points from the shell: "Open file," "Open in project," "Add to window" or whatever. 🤔

**Comment by politsin** [comment]
Во всех приложения такую задачу решают галочкой "разрешить открывать несколько копий приложения". По умолчанию галочка должна быть снята.

**Comment by aviatesk** [comment]
@smashwilson 
> I really hesitate to make any further changes without collecting more data about it. Adding a configuration option may help if there simply isn't a consensus of user expectations - but that's just punting the problem, and we still need to decide on a default, and ways of describing each option in a way that users will understand.

Great discussion.
Let me join and I want to propose one idea about crafting git commit messages:
Of course I set `[core] editor = atom --wait` in .gitconfig, but after updating to 1.36.0, every git commit or rebasing messages goes into an new window, even when I exec git command within integrated shell in an already opened window...
So I think at least a git commit message file should appear in a last focused window.
How about this idea ?

**Comment by smashwilson** [comment]
@aviatesk For that, set `[core] editor = atom --add --wait` instead. `--add` will _always_ open the new editor in the last-focused window, which is what you want anyway.

**Comment by aviatesk** [comment]
@smashwilson I tried `[core] editor = atom --add --wait` but it didn't work with v1.36.0, because the original editor didn't get returned properly but kept waiting for the other to be closed even when I'd finished editing and surely closed it.

I also confirmed with v1.35.1 `[core] editor = atom --add --wait` worked properly, even though `--add` option added the git directory to the project view, which doesn't seem nice.

**Comment by smashwilson** [comment]
>  the original editor didn't get returned properly but kept waiting for the other to be closed even when I'd finished editing and surely closed it.

I'm not sure I follow this. Can you elaborate? I did a quick check and `--add --wait` seems to be behaving as expected for me (on a Windows 10 machine at the moment).

**Comment by aviatesk** [comment]
Sure !

#### Steps to reproduce

1. Open commit message editor in Atom v.1.36.0 with .gitconfig `[core] editor = atom --add --wait` and finish editing and close the editor:
![Editing commit message](https://user-images.githubusercontent.com/40514306/56110537-da5cfd00-5f8f-11e9-8db1-bb93797cffd7.png)

2. Even though commit message editor was closed, the original commit process is not returned and thus not proceed anymore (executed from Git-Bash):
![Don't return to Git-Bash](https://user-images.githubusercontent.com/40514306/56110552-eb0d7300-5f8f-11e9-8a3a-134c6f59092b.png)

#### Expected behaviour

When we finish a Git commit message edit in Atom, it should return to the original process (e.g.: shell command)

#### Actual behaviour

Not returning to the original process.
(Before v1.36.0, it used to return to the original process)

#### Note

- The same behaviour happens with other Git commands like crafting interactive commit rebasing.
- The same thing happens when a Git commit message editor opens in Atom in safe mode

#### Versions

Atom    : 1.36.0
Electron: 2.0.18
Chrome  : 61.0.3163.100
Node    : 8.9.3

On Windows 10

**Comment by kevingrabher** [comment]
+1 on an option for working single-window.

When opening multiple files directly on an FTP Server with Transmit, there isn't really an option to open the folder as a project. This used to work fine with working in a single window. Now every file opens a new window, which makes Atom practically unusable in this workflow.

**Comment by StiftungAusNachlass** [comment]
+1 for us. No chance using atom with a ftp-programm (we use cyberduck) with this behaviour..

**Comment by Siguza** [comment]
> I really hesitate to make any further changes without collecting more data about it. Adding a configuration option may help if there simply isn't a consensus of user expectations - but that's just punting the problem, and we still need to decide on a default, and ways of describing each option in a way that users will understand.

For a start, I think we can all agree that if you open 5 files with Atom simultaneously and each gets its own window, that is counter-intuitive and wrong.  
Personally I would vote for always opening files in the last active window as well, not only because I find that the most intuitive, but also because AFAIK that's what Atom has been doing for years now.

> If an Atom window is already open with no project root folders, open new files in that Atom window. (But what do we do if there are multiple Atom windows with no project roots? What should we do when opening new folders?)

If I understood correctly, opening a folder equals setting a project root, right? If so, make it so that new folder -> new window.  
And if there are multiple windows with no project root, once again: take the last active one. Anyone having an issue with that will have had that same issue for years already anyway.

**Comment by 10686142** [comment]
Opening files in new windows doesn't work for me and probably other people as well. I can't open different files from different directories in the same window when using something like FileZilla (or even from the same directory later on), which makes my work very inefficient. Since I now have to find every window, scale it so it can fit next to another window and I have to keep doing this every time I open or close a new file

**Comment by smashwilson** [comment]
> For a start, I think we can all agree that if you open 5 files with Atom simultaneously and each gets its own window, that is counter-intuitive and wrong.

Yes, certainly. We just want to ensure that the solution to that that we decide on breaks as few other use-cases as possible.

One reason not to change our launch behavior to "always open in an existing window" is because it would have some undesirable interactions with the way we manage saved session state (unsaved buffer contents, window positioning, things like that). Currently the session is keyed by the set of normalized project root directories. If we always add to existing windows, opening a new directory would unexpectedly change the session key and make it easy to lose track of things like unsaved buffer changes. I have some ideas for improving the way that we manage sessions... but that would be a larger undertaking.

**Comment by kevingrabher** [comment]
I feel like the whole project-based workflow is counter-intuitive for a lot of people that do not load full projects into the editor but like to open files across different projects/folders/etc. in the same window, and there should be a way to have it just operate as an editor that does not look for some kind of project-structure.

**Comment by webgildor** [comment]
Hi, couldn't you just make this to be a selectable option by each user for files that are outside project folders: open files in a new window or the last focused window? Wouln't satisfy that all interests?

**Comment by btrazzini** [comment]
+1 same here!
Now every external file I open, it starts a new window. 
It's horrible, because I can not get the files in the near tabs.

**Comment by Jbrownie** [comment]
+1

**Comment by peterdthg** [comment]
This is terrible experience, not sure who made that decision but yes opening any file from Total Commander or WinSCP (even file that exist in currently opened project) opens new window, this is so frustrating that I may change back to Notepad++

IMO this is very amateurish approach to make changes like that without giving users option to control this behavior.

How do I downgrade Atom to previous version and prevent it from updating?

**Comment by stepel** [comment]
> 
> 
> This is terrible experience, not sure who made that decision but yes opening any file from Total Commander or WinSCP (even file that exist in currently opened project) opens new window, this is so frustrating that I may change back to Notepad++
> 
> IMO this is very amateurish approach to make changes like that without giving users option to control this behavior.
> 
> How do I downgrade Atom to previous version and prevent it from updating?

Exactly.. :/ 

I have downloaded atom installer from [https://github.com/atom/atom/releases ](url) and disabled auto update option.

**Comment by smashwilson** [comment]
@peterdthg:

1. Yes, if it wasn't clear from my responses above, this is a bug and we intend to fix it. As I've said, our launch handling is much more complicated than it looks, and the code that handles it is fragile, difficult to maintain, and inadequately covered by tests. I want to be sure that the solution we deliver doesn't break anything else for an equal number of other users. We're working on that presently.

2. We have a [beta channel](https://atom.io/beta) where most changes are delivered about a month before they reach the stable branch. Yes, this bug was present on the last beta, and no, nobody using beta noticed it. My working hypothesis is that beta users are more likely to be command-line users, where this isn't an issue, so it slipped through. If you're inclined to help, running beta is a great way to give us early feedback for your usage patterns.

3. As an interim solution, either adding a parent folder as a project root or downgrading to v1.35.1 is fine. You're subscribed to the issue so you should be notified when this is resolved.

**Comment by lee-dohm** [comment]
Despite @smashwilson's kind response, @peterdthg your last comment was deleted as a violation of the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) as it was insulting or derogatory. You may consider this an official warning.

**Comment by smashwilson** [comment]
This problem has been fixed and shipped with Atom 1.36.1 and 1.37.0-beta1 🚀 Thank you for your patience 🙇
